freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaInstallService.java

```diff
@@ -71,16 +81,36 @@ public void installFreeIpa(Long stackId) throws CloudbreakOrchestratorException
                 .collect(Collectors.toSet());
         FreeIpa freeIpa = freeIpaService.findByStack(stack);
 
+        if (allNodes.isEmpty()) {
+            String errorMessage = "There are no nodes to install with FreeIPA.";
+            LOGGER.error(errorMessage);
+            throw new CloudbreakOrchestratorFailedException(errorMessage);
+        }
+        GatewayConfig primaryGatewayConfig = gatewayConfigService.getPrimaryGatewayConfig(stack);
+
         SaltConfig saltConfig = new SaltConfig();
         Map<String, SaltPillarProperties> servicePillarConfig = saltConfig.getServicePillarConfig();
-        Map<String, String> freeipaPillar = Map.of("realm", freeIpa.getDomain().toUpperCase(),
-                "domain", freeIpa.getDomain(),
-                "password", freeIpa.getAdminPassword());
+        Set<Object> hosts = allNodes.stream().map(n ->
+                Map.of("ip", n.getPrivateIp(),
+                        "fqdn", n.getHostname()))
+                .collect(Collectors.toSet());
+        Map<String, Object> freeipaPillar = new HashMap<>();
+        freeipaPillar.put("realm", freeIpa.getDomain().toUpperCase());
+        freeipaPillar.put("hosts", hosts);
+        freeipaPillar.put("mode", "FIRST_INSTALL");
```

Comment:
Usually, we use grains, better to say the role field of grains, to target certain states to a particular node.  E.g.  ambari_server_standby: https://github.com/hortonworks/cloudbreak/blob/master/orchestrator-salt/src/main/resources/salt/salt/top.sls#L106

I think we can use the following roles:
freeipa_primary
freeipa_replica

Reply:
I updated it to use grains for the role assignment. I still have to see why the integration tests are failing (they don't run locally for me so I will see if I can get help on running it tomorrow).